### ISSUE TITLE
Forward declaration suggested despite specialization being available

### ISSUE THREAD

**Issue by aaronpuchert** [include-what-you-use/include-what-you-use#735]
Suppose I declare a template, but never define it, only specializations thereof. So we have template.hpp with content

```c++
template<typename>
struct T;

template<> struct T<int> { int x; };
```

and source.cpp with

```c++
#include "template.hpp"

int f(T<int>& t) { return t.x; }
```

This is a full use of the specialization, so removing the header is not suggested, but I'm told to declare the original template again:

```
source.cpp should add these lines:
template <typename> struct T;

source.cpp should remove these lines:

The full include-list for source.cpp:
#include "template.hpp"  // for T
template <typename> struct T;
```

Generally we want to repeat forward declarations, but here we have a full use of a specialization, so I think it's pretty clear that the additional declaration cannot be necessary.

**Comment by kimgr** [comment]
Hi @aaronpuchert, thanks for the report!

Based on your example alone, your conclusion makes sense. I don't know exactly how the IWYU template analysis works to say why it doesn't work as expected, but there were some improvements made in this area recently about a year ago: https://github.com/include-what-you-use/include-what-you-use/pull/578

Maybe you can use the discussion and resulting testcases to find motivation for the current behavior?

Or turn up the log level (`-Xiwyu -v7`) to see if you can figure out in which phase IWYU makes the decision.

**Comment by aaronpuchert** [comment]
> Based on your example alone, your conclusion makes sense. I don't know exactly how the IWYU template analysis works to say why it doesn't work as expected, but there were some improvements made in this area recently about a year ago: #578

If I'm not mistaken, this was talking about explicit instantiations, not specializations, of templates.

> Or turn up the log level (`-Xiwyu -v7`) to see if you can figure out in which phase IWYU makes the decision.

We have a fwd-decl use of the specialization in the parameter list of `f`, and then a full-info use of the same specialization in `t.x`. Interestingly though, on the parameter list we have a "warning: T needs a declaration, but does not provide or directly #include one." That could point to the problem.

**Comment by aaronpuchert** [comment]
This is the full log, it's not so long:

```
Setting verbose-level to 7
Search path: /usr/include/c++/7/x86_64-suse-linux/ (system)
Search path: /usr/include/c++/7/backward/ (system)
Search path: /usr/include/c++/7/ (system)
Search path: /usr/local/include/ (system)
Search path: /usr/include/ (system)
Found include: <built-in>:-1 -> source.cpp
Added to main compilation unit: source.cpp
[ Exiting to  ] source.cpp:1:1
[ #include    ] "template.hpp" (template.hpp)
Found include: source.cpp:1 -> template.hpp
Added an #include: source.cpp -> "template.hpp"
[ Exiting to  ] source.cpp:2:1
Intends-to-provide for source.cpp:
   source.cpp
   template.hpp
Intends-to-provide for template.hpp:
   template.hpp
source.cpp:3:5: (1) [ FunctionDecl ] 0x21dbe20 int f(T<int> &t)
source.cpp:3:1: (2) [ FunctionProtoType ] 0x21dbdc0 int (T<int> &)
source.cpp:3:1: (2) [ FunctionProtoTypeLoc ] int (T<int> &)
source.cpp:3:1: (3) [ BuiltinType ] 0x2179880 int
source.cpp:3:1: (3) [ BuiltinTypeLoc ] int
source.cpp:3:15: (3) [ ParmVarDecl ] 0x21dbd20 T<int> &t
source.cpp:3:7: (4) [ LValueReferenceType ] 0x21dbcc0 T<int> &
source.cpp:3:7: (4) [ LValueReferenceTypeLoc ] T<int> &
source.cpp:3:7: (5) [ TemplateSpecializationType ] 0x21dbc20 T<int>
Marked fwd-decl use of decl 0x21db8e0 T (from template.hpp:4:19) at source.cpp:3:7
source.cpp:3:7: (5, fwd decl) [ TemplateSpecializationTypeLoc ] T<int>
source.cpp:3:7: (6, fwd decl) [ TemplateName ] T
source.cpp:3:9: (6, fwd decl) [ TemplateArgumentLoc ] 0x7ffca13416f0 <int>
source.cpp:3:9: (7, fwd decl) [ BuiltinType ] 0x2179880 int
source.cpp:3:9: (7, fwd decl) [ BuiltinTypeLoc ] int
source.cpp:3:18: (2) [ CompoundStmt ] 0x21dbf88 CompoundStmt 0x21dbf88 <source.cpp:3:18, col:32>
`-ReturnStmt 0x21dbf78 <col:20, col:29>
  `-ImplicitCastExpr 0x21dbf60 <col:27, col:29> 'int' <LValueToRValue>
    `-MemberExpr 0x21dbf30 <col:27, col:29> 'int' lvalue .x 0x21dbba0
      `-DeclRefExpr 0x21dbf10 <col:27> 'T<int>':'struct T<int>' lvalue ParmVar 0x21dbd20 't' 'T<int> &'

source.cpp:3:20: (3) [ ReturnStmt ] 0x21dbf78 ReturnStmt 0x21dbf78 <source.cpp:3:20, col:29>
`-ImplicitCastExpr 0x21dbf60 <col:27, col:29> 'int' <LValueToRValue>
  `-MemberExpr 0x21dbf30 <col:27, col:29> 'int' lvalue .x 0x21dbba0
    `-DeclRefExpr 0x21dbf10 <col:27> 'T<int>':'struct T<int>' lvalue ParmVar 0x21dbd20 't' 'T<int> &'

source.cpp:3:27: (4) [ ImplicitCastExpr ] 0x21dbf60 ImplicitCastExpr 0x21dbf60 <source.cpp:3:27, col:29> 'int' <LValueToRValue>
`-MemberExpr 0x21dbf30 <col:27, col:29> 'int' lvalue .x 0x21dbba0
  `-DeclRefExpr 0x21dbf10 <col:27> 'T<int>':'struct T<int>' lvalue ParmVar 0x21dbd20 't' 'T<int> &'

source.cpp:3:29: (5) [ MemberExpr ] 0x21dbf30 MemberExpr 0x21dbf30 <source.cpp:3:27, col:29> 'int' lvalue .x 0x21dbba0
`-DeclRefExpr 0x21dbf10 <col:27> 'T<int>':'struct T<int>' lvalue ParmVar 0x21dbd20 't' 'T<int> &'

(For type T<int>):
Marked full-info use of decl 0x21db8e0 T (from template.hpp:4:19) at source.cpp:3:29
source.cpp:3:27: (6) [ DeclRefExpr ] 0x21dbf10 DeclRefExpr 0x21dbf10 <source.cpp:3:27> 'T<int>':'struct T<int>' lvalue ParmVar 0x21dbd20 't' 'T<int> &'

Marked full-info use of decl 0x21dbd20 t (from source.cpp:3:15) at source.cpp:3:27
--- Calculating IWYU violations for source.cpp ---
Ignoring use of t (source.cpp:3:27): definition is present: source.cpp:3:15
Mapped template.hpp to "template.hpp" for T (only candidate)
Ignoring full use of T (source.cpp:3:29): #including dfn from "template.hpp"
source.cpp:3:7: warning: T needs a declaration, but does not provide or directly #include one.

source.cpp should add these lines:
template <typename> struct T;

source.cpp should remove these lines:

The full include-list for source.cpp:
#include "template.hpp"  // for T
template <typename> struct T;
---
```

**Comment by kimgr** [comment]
> If I'm not mistaken, this was talking about explicit instantiations, not specializations, of templates.

I need to go over this in more detail, but I suspect the IWYU policy should be the same for explicit instantiations and explicit specializations, so maybe it's just some additional book-keeping missing for this.

**Comment by kimgr** [comment]
This patch fixes your example, and doesn't break any tests:
```diff
From eee67995e6a9690153e6a580c11da49c79184cbc Mon Sep 17 00:00:00 2001
From: Kim Grasman <kim.grasman@gmail.com>
Date: Sat, 21 Dec 2019 16:46:23 +0100
Subject: [PATCH] HAXX: Don't second-guess template specializations

---
 iwyu_output.cc | 8 ++------
 1 file changed, 2 insertions(+), 6 deletions(-)

diff --git a/iwyu_output.cc b/iwyu_output.cc
index 0d51bc9..919131f 100644
--- a/iwyu_output.cc
+++ b/iwyu_output.cc
@@ -1389,14 +1389,10 @@ void CalculateIwyuForForwardDeclareUse(
 
   // If this record is defined in one of the desired_includes, mark that
   // fact.  Also if it's defined in one of the actual_includes.
-  const NamedDecl* dfn = GetDefinitionForClass(use->decl());
-  // If we are, ourselves, a template specialization, then the definition
-  // we use is not the definition of the specialization (that's us), but
-  // the definition of the template we're specializing.
-  if (spec_decl && dfn == spec_decl)
-    dfn = GetDefinitionForClass(spec_decl->getSpecializedTemplate());
   bool dfn_is_in_desired_includes = false;
   bool dfn_is_in_actual_includes = false;
+
+  const NamedDecl* dfn = GetDefinitionForClass(use->decl());
   if (dfn) {
     vector<string> headers
       = GlobalIncludePicker().GetCandidateHeadersForFilepathIncludedFrom(
-- 
2.20.1
```

Could you give it a try in the real world and see if it seems to help?

**Comment by aaronpuchert** [comment]
> I need to go over this in more detail, but I suspect the IWYU policy should be the same for explicit instantiations and explicit specializations, so maybe it's just some additional book-keeping missing for this.

I think they are somewhat different, but you're right that in both cases we can assume a forward declaration of the original template is already available, because otherwise neither a specialization nor an explicit instantiation declaration would be well-formed.

> This patch fixes your example, and doesn't break any tests:
> [...]
> Could you give it a try in the real world and see if it seems to help?

Thanks! I guess I'll have a look after the holidays, I hope you don't mind. But given that it fixes my test case and doesn't break anything else, I have a good feeling about it. Also the comment that you removed seems somewhat odd, since a specialization requires only a declaration, not a definition of the original template.

**Comment by aaronpuchert** [comment]
Finally got around to test the patch. There are no more superfluous suggestions in any of the instances where I have observed this issue. So this is good to go from my side.

**Comment by kimgr** [comment]
This was fixed by PR #765, closing.
